packages/plugins/sasl/index.js

```diff
@@ -57,12 +57,13 @@ module.exports = plugin('sasl', {
 
     return Promise.resolve(this.getMechanism(offered, usable, available, features)).then(mech => {
       this.mech = mech
-      return Promise.resolve(this.handleMechanism(mech, features))
+      return this.handleMechanism(mech, features)
     })
   },
 
   handleMechanism(mech, features) {
-    return Promise.resolve(this.getCrendentials(mech, features)).then((username, password) => {
+    this.entity._status('authenticate')
+    return this.entity.delegate('authenticate', (username, password) => {
       return this.authenticate(mech, {username, password}, features)
```

Comment:
Shouldn't we just pass this `{username, password}` object from the handler, to make details of mechanism opaque to SASL plugin? Other mechanisms, like X-OAUTH2 will require different credentials than username+password.

(We also need a way to pass the details of the mechanism from the mechanism plugin to the authenticate handler. i.e. challenge string in case of challenge/response based mechanisms.)